### ISSUE TITLE
Knockback + flying mount

### ISSUE THREAD

**Issue by Bluedragoness** [TrinityCore/TrinityCore#6099]
http://www.wowwiki.com/Knock_back

Currently when you are hit by a knock back effect whilst on a flying mount, you are not knocked off of your mount but your flying mount becomes crippled. You end up falling and taking fall damage instead of being able to get back to flying regardless of the height and distance from the ground. This can ultimately lead to death if high enough off the ground.

As of 3.3.0
" Patch 3.3.0 (2009-12-08): Knockback effects no longer dismount players. If on a flying mount, you will be knocked back a short distance before being able to resume flying. This change is most likely due to some players abusing their knock back abilities on flagged players leading them to fall to their death."

EDIT: Revision f8e9dedbe5e4e0d311

**Comment by Rushor** [comment]
```c++
diff --git a/src/server/game/Spells/SpellEffects.cpp b/src/server/game/Spells/SpellEffects.cpp
index 875a3ff..02683b7 100644
--- a/src/server/game/Spells/SpellEffects.cpp
+++ b/src/server/game/Spells/SpellEffects.cpp
@@ -5104,6 +5104,10 @@ void Spell::EffectKnockBack(SpellEffIndex effIndex)
     if (unitTarget->HasUnitState(UNIT_STATE_ROOT))
         return;
 
+    // No Effect on Flying Player
+    if (unitTarget->ToPlayer()->IsFlying())
+        return;
+
     // Instantly interrupt non melee spells being cast
     if (unitTarget->IsNonMeleeSpellCast(true))
         unitTarget->InterruptNonMeleeSpells(true);
```

i tried this in the spelleffect knockback

unfortunaltly this 'fix' caused serveral crashes. cannot find the crashlog anymore and you cannot reproduce this if you test it manually with 2 chars. someone can se the issue here?

**Comment by Nyeriah** [comment]
Is the unit always granted to be a player?

**Comment by dufernst** [comment]
Indeed if the unitTarget is not a player `unitTarget->ToPlayer()` returns NULL.
If you then call `->IsFlying()` you get a null pointer exception.

**Comment by robinsch** [comment]
´´´diff
From c05e299384872e92a1bc57dbe2d37a3d757014fa Mon Sep 17 00:00:00 2001
From: robinsch <robin.schriever.hude@web.de>
Date: Wed, 27 May 2015 21:52:09 +0200
Subject: [PATCH] #6099

---
 src/server/game/Entities/Creature/Creature.cpp | 2 +-
 src/server/game/Entities/Creature/Creature.h   | 2 +-
 src/server/game/Entities/Player/Player.cpp     | 4 ++--
 src/server/game/Entities/Player/Player.h       | 2 +-
 src/server/game/Entities/Unit/Unit.cpp         | 7 +++++--
 src/server/game/Entities/Unit/Unit.h           | 2 +-
 6 files changed, 11 insertions(+), 8 deletions(-)

diff --git a/src/server/game/Entities/Creature/Creature.cpp b/src/server/game/Entities/Creature/Creature.cpp
index 7f00fc7..d04691d 100644
--- a/src/server/game/Entities/Creature/Creature.cpp
+++ b/src/server/game/Entities/Creature/Creature.cpp
@@ -2415,7 +2415,7 @@ bool Creature::SetSwim(bool enable)
     return true;
 }
 
-bool Creature::SetCanFly(bool enable)
+bool Creature::SetCanFly(bool enable, bool packetOnly /* = false */)
 {
     if (!Unit::SetCanFly(enable))
         return false;
diff --git a/src/server/game/Entities/Creature/Creature.h b/src/server/game/Entities/Creature/Creature.h
index 6f82d5d..764a57d 100644
--- a/src/server/game/Entities/Creature/Creature.h
+++ b/src/server/game/Entities/Creature/Creature.h
@@ -472,7 +472,7 @@ class Creature : public Unit, public GridObject<Creature>, public MapObject
         bool SetWalk(bool enable) override;
         bool SetDisableGravity(bool disable, bool packetOnly = false) override;
         bool SetSwim(bool enable) override;
-        bool SetCanFly(bool enable) override;
+        bool SetCanFly(bool enable, bool packetOnly = false) override;
         bool SetWaterWalking(bool enable, bool packetOnly = false) override;
         bool SetFeatherFall(bool enable, bool packetOnly = false) override;
         bool SetHover(bool enable, bool packetOnly = false) override;
diff --git a/src/server/game/Entities/Player/Player.cpp b/src/server/game/Entities/Player/Player.cpp
index 030c2a2..275dbf9 100644
--- a/src/server/game/Entities/Player/Player.cpp
+++ b/src/server/game/Entities/Player/Player.cpp
@@ -26251,9 +26251,9 @@ bool Player::SetDisableGravity(bool disable, bool packetOnly /*= false*/)
     return true;
 }
 
-bool Player::SetCanFly(bool apply)
+bool Player::SetCanFly(bool apply, bool packetOnly /*= false*/)
 {
-    if (!Unit::SetCanFly(apply))
+    if (!packetOnly && !Unit::SetCanFly(apply))
         return false;
 
     WorldPacket data(apply ? SMSG_MOVE_SET_CAN_FLY : SMSG_MOVE_UNSET_CAN_FLY, 12);
diff --git a/src/server/game/Entities/Player/Player.h b/src/server/game/Entities/Player/Player.h
index af640e4..79fc499 100644
--- a/src/server/game/Entities/Player/Player.h
+++ b/src/server/game/Entities/Player/Player.h
@@ -2301,7 +2301,7 @@ class Player : public Unit, public GridObject<Player>
         void RemoveFromWhisperWhiteList(ObjectGuid guid) { WhisperList.remove(guid); }
 
         bool SetDisableGravity(bool disable, bool packetOnly /* = false */) override;
-        bool SetCanFly(bool apply) override;
+        bool SetCanFly(bool apply, bool packetOnly = false) override;
         bool SetWaterWalking(bool apply, bool packetOnly = false) override;
         bool SetFeatherFall(bool apply, bool packetOnly = false) override;
         bool SetHover(bool enable, bool packetOnly = false) override;
diff --git a/src/server/game/Entities/Unit/Unit.cpp b/src/server/game/Entities/Unit/Unit.cpp
index d57ae82..c6faf84 100644
--- a/src/server/game/Entities/Unit/Unit.cpp
+++ b/src/server/game/Entities/Unit/Unit.cpp
@@ -16440,7 +16440,7 @@ void Unit::KnockbackFrom(float x, float y, float speedXY, float speedZ)
         float vcos, vsin;
         GetSinCos(x, y, vsin, vcos);
 
-        WorldPacket data(SMSG_MOVE_KNOCK_BACK, (8+4+4+4+4+4));
+        WorldPacket data(SMSG_MOVE_KNOCK_BACK, (8 + 4 + 4 + 4 + 4 + 4));
         data << GetPackGUID();
         data << uint32(0);                                      // counter
         data << float(vcos);                                    // x direction
@@ -16449,6 +16449,9 @@ void Unit::KnockbackFrom(float x, float y, float speedXY, float speedZ)
         data << float(-speedZ);                                 // Z Movement speed (vertical)
 
         player->GetSession()->SendPacket(&data);
+
+        if (player->HasAuraType(SPELL_AURA_MOD_INCREASE_MOUNTED_FLIGHT_SPEED) || player->HasAuraType(SPELL_AURA_FLY))
+            player->SetCanFly(true, true);
     }
 }
 
@@ -17519,7 +17522,7 @@ bool Unit::SetSwim(bool enable)
     return true;
 }
 
-bool Unit::SetCanFly(bool enable)
+bool Unit::SetCanFly(bool enable, bool /*packetOnly = false */)
 {
     if (enable == HasUnitMovementFlag(MOVEMENTFLAG_CAN_FLY))
         return false;
diff --git a/src/server/game/Entities/Unit/Unit.h b/src/server/game/Entities/Unit/Unit.h
index 6b47ee9..1ae377b 100644
--- a/src/server/game/Entities/Unit/Unit.h
+++ b/src/server/game/Entities/Unit/Unit.h
@@ -1585,7 +1585,7 @@ class Unit : public WorldObject
         virtual bool SetWalk(bool enable);
         virtual bool SetDisableGravity(bool disable, bool packetOnly = false);
         virtual bool SetSwim(bool enable);
-        virtual bool SetCanFly(bool enable);
+        virtual bool SetCanFly(bool enable, bool packetOnly = false);
         virtual bool SetWaterWalking(bool enable, bool packetOnly = false);
         virtual bool SetFeatherFall(bool enable, bool packetOnly = false);
         virtual bool SetHover(bool enable, bool packetOnly = false);
-- 
1.9.5.msysgit.1

**Comment by tkrokli** [comment]
Would you like a PR on https://github.com/TrinityCore/TrinityCore/issues/6099#issuecomment-106052609 or does it need more testing to confirm that it is safe to use?

**Comment by robinsch** [comment]
Should be save to use. You can create a PR for it. Thanks!

**Comment by SmythX** [comment]
how to add this fix to 3.3.5
